src/Diagnostics.DataProviders/KustoHeartBeatService/KustoHeartBeatService.cs

```diff
@@ -0,0 +1,176 @@
+        }
+
+        public KustoHeartBeatService(KustoDataProviderConfiguration configuration)
+        {
+            _configuration = configuration;
+            Initialize();
+        }
+
+        public string GetClusterNameFromStamp(string stampName)
+        {
+            string kustoClusterName = null;
+            string appserviceRegion = ParseRegionFromStamp(stampName);
+
+            if (!_configuration.RegionSpecificClusterNameCollection.TryGetValue(appserviceRegion.ToLower(), out kustoClusterName))
+            {
+                if (!_configuration.RegionSpecificClusterNameCollection.TryGetValue("*", out kustoClusterName))
+                {
+                    throw new KeyNotFoundException(String.Format("Kusto Cluster Name not found for Region : {0}", appserviceRegion.ToLower()));
+                }
+            }
+
+            if (_heartbeats.ContainsKey(kustoClusterName))
+            {
+                if (!_heartbeats[kustoClusterName].UsePrimaryCluster)
+                {
+                    kustoClusterName = _heartbeats[kustoClusterName].FailoverCluster;
+                }
+            }
+
+            return kustoClusterName;
+        }
+
+        private static string ParseRegionFromStamp(string stampName)
+        {
+            if (string.IsNullOrWhiteSpace(stampName))
+            {
+                throw new ArgumentNullException(nameof(stampName));
+            }
+
+            var stampParts = stampName.Split(new char[] { '-' });
+            if (stampParts.Any() && stampParts.Length >= 3)
+            {
+                return stampParts[2];
+            }
+
+            //return * for private stamps if no prod stamps are found
+            return "*";
+        }
+    }
+
+    public class KustoHeartBeat
+    {
+        public string PrimaryCluster { get; }
+        public string FailoverCluster { get; }
+        public bool UsePrimaryCluster { get; private set; }
+        private int _ConsecutiveFailureCount = 0;
+        private int _ConsecutiveSuccessCount = 0;
+        private KustoDataProvider _kustoDataProvider;
+
+        public KustoHeartBeat(string primaryCluster, string failoverCluster, KustoDataProvider kustoDataProvider)
+        {
+            PrimaryCluster = primaryCluster;
+            FailoverCluster = failoverCluster;
+            UsePrimaryCluster = true;
+            _kustoDataProvider = kustoDataProvider;
+        }
+
+        public async void RunHeartBeatThread()
+        {
+            while (true) {
+                bool heartBeatSuccess = false;
+                Exception exceptionForLog = null;
+                string requestID = Guid.NewGuid().ToString();
+                try
+                {
+                    // Run kusto query with  10 second time out
+                    string query = "RoleInstanceHeartbeat | where TIMESTAMP >= ago(30m) | take 1";
+                    var result = await _kustoDataProvider.ExecuteQueryDirect(query, PrimaryCluster, 10, requestID);
```

Comment:
Let's make the timeout configurable in the `KustoDataProviderConfiguration` so that we don't have to recompile the source code when we change the timeout.